### PR TITLE
Add: Added a new CVE scan method with the new CPE matching.

### PR DESCRIPTION
**Pull request by jhelmold**
## What
Added a new CVE scan method with the new CPE matching.

## Why
These new functions are needed for the new CVE scans with JSON match rules for CPE matching.

## Reference
GEA-626

## Checklist

Tested manually on my local development system.
- [ ] Tests

### CODE REVIEW

File: src/manage_sql.c [hunk 1]
```diff
@@ -23086,6 +23288,8 @@ results_extra_where (int trash, report_t report, const gchar* host,
 
   min_qod_clause = where_qod (min_qod);
 
+  g_message ("PROTO1: %s", levels);
```
Comment (mattmundell) [inline src/manage_sql.c:23291]: Looks like stray tracing.

File: src/manage.c [hunk 2]
```diff
@@ -3106,6 +3109,138 @@ set_scanner_connection_retry (int new_retry)
...
+    {
+      iterator_t cpe_host_details_products;
+      gchar *range_fs_cpe;
+      gchar *range_uri_product;
+      gchar *vsi, *vse, *vei, *vee;
+      range_fs_cpe = vsi = vse = vei = vee = NULL;
+      range_fs_cpe = g_strdup (cpe_match_range_iterator_cpe (&cpe_match_ranges));
+      vsi = (gchar*) cpe_match_range_iterator_version_start_incl(&cpe_match_ranges);
+      if (vsi != NULL)
+        vsi = g_strdup (cpe_match_range_iterator_version_start_incl(&cpe_match_ranges));
```
Comment (mattmundell) [inline src/manage.c:3192]:
```suggestion
        vsi = g_strdup (cpe_match_range_iterator_version_start_incl (&cpe_match_ranges));
```
and more below

**Review by mattmundell — CHANGES_REQUESTED** [review]
Looks like a good change.

File: src/manage.c [hunk 3]
```diff
@@ -3161,89 +3296,220 @@ cve_scan_host (task_t task, report_t report, gvm_host_t *gvm_host)
...
-          while (next (&prognosis))
+
+          gboolean use_json = FALSE;
+          if (sql_int64_0 ("SELECT count(1) FROM information_schema.tables"
+                           " WHERE table_schema = 'scap'"
+                           " AND table_name = 'cpe_match_nodes';") > 0)
+            use_json = TRUE;
+
+          if (use_json)
             {
```
[the remark this answers is not in the document]
Comment (mattmundell) [reply src/manage.c:3445]: Nice, thanks.

File: src/manage_sql.c [hunk 4]
```diff
@@ -22357,6 +22549,16 @@ where_levels_auto (const char *levels, const char *new_severity_sql)
       g_string_append (levels_sql, ", 'false'");
       count++;
     }
+  if (strchr (levels, 'e'))
+    {
+      g_string_append (levels_sql, ", 'error'");
+      count++;
+    }
+  if (strchr (levels, 'd'))
```
[the remark this answers is not in the document]
Comment (mattmundell) [reply src/manage_sql.c:22557]: Thanks for removing 'd', but I really think 'e' should be in a separate PR with an explanation.

Has this additional level been agreed on? Should it be added to the GMP doc? Does it affect delta reports? Should `manage_report_filter_controls` mention it? How about a `<filter>Error</filter>` in `print_report_xml_start`? Does the gsad levels validator need adjustment? Does GSA need a checkbox for it? It definitely should be mentioned in the docs of `where_levels_auto`.

Note that there's already a separate mechanism for returning errors in reports, see `print_report_errors_xml`.

File: src/manage_sql_secinfo.c [hunk 5]
```diff
@@ -2703,7 +2703,9 @@ json_object_item_double (cJSON *object, char *key, double fallback)
  * @brief Save the node of a cve match rule tree.
  *
  * @param[in]  parent_id  The parent_id of the node. If this value is 0,
- *                        this node is the root of the tree.
+ *                        the node is the root of the tree.
+ * @param[in]  root_id    The id of the root of the tree. If this value
```
[the remark this answers is not in the document]
Comment (mattmundell) [reply src/manage_sql_secinfo.c]: still missing.

**Comment by mattmundell** [comment]
@jhelmold I replied above, in case github is hiding them.

File: src/manage_sql.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (jhelmold) [reply src/manage_sql.c:22557]: I discussed this topic with Timo. We decided not to support any option like "d" or "e" for the display of errors. So I removed also the "e" option.

**Review by mattmundell — APPROVED** [review]

**Pull request merged**